pkg/loki/loki.go

```diff
@@ -0,0 +1,89 @@
+package loki
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/grafana/loki/pkg/logcli/client"
+	"github.com/grafana/loki/pkg/loghttp"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/pingcap/errors"
+)
+
+const (
+	equalMatcher    = "|="
+	regexMatcher    = "|~"
+)
+
+type lokiClient struct {
+	cli       *client.Client
+	startTime time.Time
+}
+
+// NewLokiClient creates a client to query loki.
+func NewLokiClient(address, username, password string) *lokiClient {
+	return &lokiClient{
+		cli: &client.Client{
+			Address:  address,
+			Username: username,
+			Password: password,
+		},
+		startTime: time.Now(),
+	}
+}
+
+// FetchContainerLogs gets container logs from loki API. ns and containerName are used
+// to match the specific logs in loki. match is a string which you want to query
+// from loki, you can set isRegex to true to make it to be a regex match. nonMatch
+//  is a set of strings you don't want to match. lookbackTime is the query duration.
+func (c *lokiClient) FetchContainerLogs(ns, containerName, match string, nonMatch []string, lookbackTime time.Duration, isRegex bool) ([]string, int, error) {
+	if ns == "" {
+		return nil, 0, errors.New("namespace must be set")
+	}
+
+	if containerName == "" {
+		return nil, 0, errors.New("container name must be set")
+	}
+
+	if match == "" {
+		return nil, 0, errors.New("match query must be set")
+	}
+
+	var op string
+	if isRegex {
+		op = regexMatcher
+	} else {
+		op = equalMatcher
+	}
+
+	// Format the query to loki.
+	query := fmt.Sprintf(`{container_name="%s", namespace="%s"} %s"%s"`,
+		containerName, ns, op, match)
+
+	var nonEqual string
+	for _, v := range nonMatch {
+		nonEqual += fmt.Sprintf(` !="%s"`, v)
+	}
+	query += nonEqual
+
+	queryTo := time.Now()
+	queryFrom := queryTo.Add(-lookbackTime)
+	if queryFrom.Before(c.startTime) {
+		queryFrom = c.startTime
+	}
```

Comment:
Just an advice, if `lookbackTime` is set to 0, it will query all logs in the history, unless it query between the given duration.

Reply:
What do you mean by `query all logs`? But thanks, I checked the code in loki and I really need to handle this case. If `starttime == endtime`, it will return an error.

Reply:
I was considering setting `startTime` to 0 for all logs, but here we should avoid to get logs from previous cases. Maybe we can just handle the case `queryFrom == queryTo` and return empty result for now.

Reply:
Or maybe we can change the method signature? pass startTime and endTime maybe more convenient?

Reply:
Pass startTime and endTime is ok. But the startTime check is required unless we can find a way to just query the cluster we want.